### PR TITLE
chore: Add Dockerfile and publish workflow

### PR DESCRIPTION
**Pull request by poppoerika**
momento-proxy Docker image is tested on an EC2 instance and I was able to set a key/value pair.
```
[root@ip-172-31-53-115 bin]# telnet 0.0.0.0 11211
Trying 0.0.0.0...
Connected to 0.0.0.0.
Escape character is '^]'.
set erika 0 0 3
eri 
STORED
```

Every time a pr is merged to `main`, a new Docker image is built and pushed to the `gomomento/momento-proxy` repo.
When users run `docker run`, they need to provide an env variable `MOMENTO_AUTHENTICATION` (Momento auth token).
Also updated `momento-proxy.md` to include how to run the momento-proxy Docker image.

### CODE REVIEW

File: src/rust/protocol/memcache/src/memcache/wire/request/parse.rs [hunk 1]
```diff
@@ -198,7 +198,7 @@ fn parse_get(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
                     } else {
                         #[cfg(feature = "stats")]
                         GET_CARDINALITY.increment(PreciseInstant::now(), keys.len() as _, 1);
-                        
+
```
Comment (eaddingtonwhite) [inline src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:201]: Id just revert this file so your commit isnt modifying any rust files just so easier to review when we contribute down.

File: src/rust/proxy/momento/README.md [hunk 2]
```diff
@@ -41,18 +41,34 @@ use with the proxy.
...
-```cargo run --release --bin momento_proxy -- path/to/config.toml```
+`cargo run --release --bin momento_proxy -- path/to/config.toml`
 
 Your application can now connect to the proxy on the configured port(s) to send
 requests to the corresponding Momento cache(s).
 
 The resulting binary will be `target/release/momento_proxy` and it can be copied
 to a standard system path (eg: `/usr/local/bin`).
+
+### momento-proxy Docker image
```
Comment (eaddingtonwhite) [inline src/rust/proxy/momento/README.md:60]: we should have instructions here for how to build, run, and test container locally as well

File: src/rust/protocol/memcache/src/memcache/wire/request/parse.rs [hunk 3]
```diff
@@ -198,7 +198,6 @@ fn parse_get(buffer: &[u8]) -> Result<ParseOk<MemcacheRequest>, ParseError> {
                     } else {
                         #[cfg(feature = "stats")]
                         GET_CARDINALITY.increment(PreciseInstant::now(), keys.len() as _, 1);
-
```
Comment (eaddingtonwhite) [inline src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:201]: I would revert this line

File: src/rust/proxy/momento/README.md [hunk 4]
```diff
@@ -41,18 +41,66 @@ use with the proxy.
...
+- Building a new momento-proxy image:
+
+```
+docker build --tag momento-proxy .
+```
+
+- Running the newly built image with the default config:
+
+```
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN>
```
Comment (eaddingtonwhite) [inline src/rust/proxy/momento/README.md:87]: are we missing image name in this? Doesnt `momento-proxy` need to be in command somewhere?

File: src/rust/protocol/memcache/src/memcache/wire/request/parse.rs (hunk 3 above)
Comment (poppoerika) [reply src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:201]: Ah... I think VSCode formatted it on save.

File: src/rust/protocol/memcache/src/memcache/wire/request/parse.rs (hunk 1 above)
Comment (poppoerika) [reply src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:201]: @eaddingtonwhite https://github.com/brayniac/pelikan/blob/momento-proxy/src/rust/protocol/memcache/src/memcache/wire/request/parse.rs#L201
The original file has a new line on 201. So added it back.

File: src/rust/proxy/momento/README.md [hunk 5]
```diff
@@ -41,18 +41,66 @@ use with the proxy.
...
 The resulting binary will be `target/release/momento_proxy` and it can be copied
 to a standard system path (eg: `/usr/local/bin`).
+
+### momento-proxy Docker image
+
+Running the momento-proxy image to run the Momento proxy.
+
+```
+docker pull gomomento/momento-proxy
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> momento-proxy
```
Comment (eaddingtonwhite) [inline src/rust/proxy/momento/README.md:66]: does this need to be `gomomento/momento-proxy` for image to run? vs  just `momento-proxy` below for local y built version?

Comment (poppoerika) [reply src/rust/proxy/momento/README.md:66]: Yup, I confirmed that it needs to be `gomomento/momento-proxy` to pull the image from our DockerHub repo. 
I think we might be able to do `momento-proxy` with a paid subscription (business) to DockerHub.

Comment (eaddingtonwhite) [reply src/rust/proxy/momento/README.md:66]: So should it be like this now?
```suggestion
docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> gomomento/momento-proxy
```

File: src/rust/proxy/momento/README.md [hunk 6]
```diff
@@ -41,18 +41,66 @@ use with the proxy.
...
+```
+
+By default, [this configuration](../../../../config/momento_proxy.toml) is used for the Momento proxy.
+To set your own, please provide an env variable `CONFIG` when running a container.
+
+```
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> -e CONFIG=<YOUR_CONFIG_FILE> momento-proxy
```
Comment (eaddingtonwhite) [inline src/rust/proxy/momento/README.md:73]: Same here should this be like:
```suggestion
docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> -e CONFIG=<YOUR_CONFIG_FILE> gomomento/momento-proxy
```

File: src/rust/proxy/momento/README.md [hunk 7]
```diff
@@ -41,18 +41,66 @@ use with the proxy.
...
 
 Your application can now connect to the proxy on the configured port(s) to send
 requests to the corresponding Momento cache(s).
 
 The resulting binary will be `target/release/momento_proxy` and it can be copied
 to a standard system path (eg: `/usr/local/bin`).
+
+### momento-proxy Docker image
+
+Running the momento-proxy image to run the Momento proxy.
```
Comment (eaddingtonwhite) [inline src/rust/proxy/momento/README.md:62]: Maybe reword lik:
```suggestion
You can run the `momento-proxy` container by pulling it from [Momento's](https://momentohq.com/) [docker hub registry](https://hub.docker.com/u/gomomento) with following commands. 
```

File: src/rust/protocol/memcache/src/memcache/wire/request/parse.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (poppoerika) [reply src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:201]: I think I added it back? 
I see a new line in my file...
image

[the remark this answers is not in the document]
Comment (eaddingtonwhite) [reply src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:201]: Its more then new line though its a new line plus some white space 😬. You can see in diff my suggestion block ☝️

[the remark this answers is not in the document]
Comment (poppoerika) [reply src/rust/protocol/memcache/src/memcache/wire/request/parse.rs:201]: 💡

**Review by eaddingtonwhite — APPROVED** [review]
@poppoerika This is looking good to me nice work! Lets get it merged and can do final verification again in our main `momento` branch make sure image gets published properly. Then can work to get contribute branch down into Bryan's fork. 

Maybe have @danielamiao take one more pass on content quick just to get one more pair eyes.

File: src/rust/proxy/momento/README.md [hunk 8]
```diff
@@ -41,18 +41,66 @@ use with the proxy.
...
+
+You can run the `momento-proxy` container by pulling it from [Momento's](https://momentohq.com/) [docker hub registry](https://hub.docker.com/u/gomomento) with following commands.
+
+```
+docker pull gomomento/momento-proxy
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> gomomento/momento-proxy
+```
+
+By default, [this configuration](../../../../config/momento_proxy.toml) is used for the Momento proxy.
+To set your own, please provide an env variable `CONFIG` when running a container.
```
Comment (danielamiao) [inline src/rust/proxy/momento/README.md:70]: My Dockerfile knowledge is limited, but have you tried this with a custom configuration file on EC2 instance? In the above Dockerfile I see a line where you are setting the default `CONFIG` value, and a line for copying the default config `momento_proxy.toml`, then a run command line with `CONFIG` env. However, if I supply a custom CONFIG path, how does that get copied into the Docker container?

Comment (poppoerika) [reply src/rust/proxy/momento/README.md:70]: I think `-v` needs to be passed with `docker run` to mount a volume to a Docker container. 
I'll test it out first and update the README accordingly.

**Comment by poppoerika** [comment]
@eaddingtonwhite @danielamiao 
Updated the Dockerfile a bit so that we can mount a host machine's config directory to a container's config directory to be able to use a custom config file passed by users.
Also updated README to reflect this change.
Tested it on an EC2 instance and was able to use my custom config file to run.

File: src/rust/proxy/momento/README.md [hunk 9]
```diff
@@ -41,18 +41,66 @@ use with the proxy.
...
+- Running the newly built image with the default config:
+
+```
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> momento-proxy
+```
+
+- Running the newly built image with your custom config:
+
+```
+docker run -d -p 11211:11211 -p 9999:9999 -e MOMENTO_AUTHENTICATION=<YOUR_MOMENTO_TOKEN> -e CONFIG=<YOUR_CONFIG_FILE> momento-proxy
```
Comment (eaddingtonwhite) [inline src/rust/proxy/momento/README.md:93]: does this need to be updated to have a `-v` flag passed for mounting config file?

**Review by eaddingtonwhite — APPROVED** [review]
Nice work @poppoerika
